kernel/comps/framebuffer/src/console.rs

```diff
@@ -0,0 +1,199 @@
+pub static CONSOLE_NAME: &str = "Framebuffer-Console";
+
+pub static FRAMEBUFFER_CONSOLE: Once<Arc<FramebufferConsole>> = Once::new();
+
+pub(crate) fn init() {
+    let Some(fb) = FRAMEBUFFER.get() else {
+        log::warn!("Framebuffer not initialized");
+        return;
+    };
+
+    FRAMEBUFFER_CONSOLE.call_once(|| Arc::new(FramebufferConsole::new(fb.clone())));
+}
+
+impl AnyConsoleDevice for FramebufferConsole {
+    fn send(&self, buf: &[u8]) {
+        self.send_buf(buf);
+    }
+
+    fn register_callback(&self, _: &'static ConsoleCallback) {
+        // Unsupported, do nothing.
+    }
+}
+
+impl FramebufferConsole {
+    /// Creates a new framebuffer console.
+    pub fn new(framebuffer: Arc<FrameBuffer>) -> Self {
+        let bytes = alloc::vec![0u8; framebuffer.size()];
+        Self {
+            state: SpinLock::new(ConsoleState {
+                enabled: true,
+                x_pos: 0,
+                y_pos: 0,
+                fg_color: Pixel::WHITE,
+                bg_color: Pixel::BLACK,
+                bytes,
+                backend: framebuffer,
+            }),
+        }
+    }
+
+    /// Returns whether the console is enabled.
+    pub fn is_enabled(&self) -> bool {
+        self.state.lock().enabled
+    }
+
+    /// Enables the console.
+    pub fn enable(&self) {
+        self.state.lock().enabled = true;
+    }
+
+    /// Disables the console.
+    pub fn disable(&self) {
+        self.state.lock().enabled = false;
+    }
+
+    /// Returns the current cursor position.
+    pub fn cursor(&self) -> (usize, usize) {
+        let state = self.state.lock();
+        (state.x_pos, state.y_pos)
+    }
+
+    /// Sets the cursor position.
+    pub fn set_cursor(&self, x: usize, y: usize) {
+        let mut state = self.state.lock();
+        if x > state.backend.width() - 8 || y > state.backend.height() - 8 {
+            log::warn!("Invalid framebuffer cursor position: ({}, {})", x, y);
+            return;
+        }
+        state.x_pos = x;
+        state.y_pos = y;
+    }
+
+    /// Returns the foreground color.
+    pub fn fg_color(&self) -> Pixel {
+        self.state.lock().fg_color
+    }
+
+    /// Sets the foreground color.
+    pub fn set_fg_color(&self, val: Pixel) {
+        self.state.lock().fg_color = val;
+    }
+
+    /// Returns the background color.
+    pub fn bg_color(&self) -> Pixel {
+        self.state.lock().bg_color
+    }
+
+    /// Sets the background color.
+    pub fn set_bg_color(&self, val: Pixel) {
+        self.state.lock().bg_color = val;
+    }
+
+    /// Sends a single character to be drawn on the framebuffer.
+    pub fn send_char(&self, c: char) {
+        let mut state = self.state.lock();
+        if !state.enabled {
+            return;
+        }
+
+        if c == '\n' {
+            state.newline();
+            return;
+        } else if c == '\r' {
+            state.carriage_return();
+            return;
+        }
+
+        if state.x_pos >= state.backend.width() {
```

Comment:
```suggestion
        if state.x_pos + 8 > state.backend.width() {
```
This will work if `width() % 8 != 0`.